### PR TITLE
MICROPROFILE #4490 - Referenced component name in OpenApi 'paths' sec…

### PR DESCRIPTION
**Pull request by Pscheidl**
…tion use name from the Schema annotation

# Description
This is a bug fix of https://github.com/payara/Payara/issues/4490

# Important Info

# Testing
Created test class `OASchemaReferenceNameTest`. Tests whether the name is picked up from the Schema annotation, if present and the `name ` is not empty (default is empty). Otherwise it will use simple name of the class - which is in line with how the "components" schemas behave.

### Testing Environment
`JDK 1.8.0_241`

# Documentation

# Notes for Reviewers
`OASchemaReferenceNameTest` and its `schemaComponentReferenceNaming` method. The test class is self-contained - everything required for this particular test case is there.

### CODE REVIEW

**Comment by Pscheidl** [comment]
There are some unnecessary formattings done - I'll fix them.

**Comment by Pscheidl** [comment]
Removed unnecessary formatting changes. Ready to be reviewed I believe.

**Comment by smillidge** [comment]
Fantastic. Thanks for the contribution Pavel.
I can't find a CLA on record can you print and scan as described in https://github.com/payara/Payara/blob/master/.github/CONTRIBUTING.md

**Comment by Pscheidl** [comment]
> Fantastic. Thanks for the contribution Pavel.
> I can't find a CLA on record can you print and scan as described in https://github.com/payara/Payara/blob/master/.github/CONTRIBUTING.md

My apologies, I overlooked that. Of course, I'll happily send it late this evening.

**Comment by Pscheidl** [comment]
Sent.

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java [hunk 1]
```diff
@@ -1031,11 +1031,18 @@ private boolean insertObjectReference(ApiContext context, Reference<?> referee,
...
-        referee.setRef(referenceClass.getSimpleName());
+        if (schema == null || schema.name().isEmpty()) {
+            referee.setRef(referenceClass.getSimpleName());
+        } else {
+            referee.setRef(schema.name());
+        }
 
         // Create the schema
         visitSchema(AnnotationInfo.valueOf(referenceClass).getAnnotation(Schema.class), referenceClass, context);
+        visitSchema(schema, referenceClass, context);
```
Comment (rdebusscher) [inline appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: The `visitSchema()` is now called twice with the same parameters.

File: appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    @GET
+    @Path("/address")
+    @APIResponse(description = "get address")
+    public AddressDto getAddress() {
+        return new AddressDto();
+    }
+
+    @Test
+    public void schemaComponentReferenceNaming() {
```
Comment (rdebusscher) [inline appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java]: A second test where no Schema annotation is used and checked if the class name is used is also required.

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java (hunk 1 above)
Comment (Pscheidl) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: Thank you. That line is completely in vain there. Removed in https://github.com/payara/Payara/pull/4494/commits/369794a7181575d204db98aaad315e623ae661d8

**Comment by rdebusscher** [comment]
Jenkins test please

File: appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) [2019] Payara Foundation and/or its affiliates. All rights reserved.
```
Comment (jbee) [inline appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java]: 2020

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java [hunk 4]
```diff
@@ -1031,11 +1031,17 @@ private boolean insertObjectReference(ApiContext context, Reference<?> referee,
...
 
+        final Schema schema = AnnotationInfo.valueOf(referenceClass).getAnnotation(Schema.class);
+
         // Set the reference name
-        referee.setRef(referenceClass.getSimpleName());
+        if (schema == null || schema.name().isEmpty()) {
+            referee.setRef(referenceClass.getSimpleName());
+        } else {
+            referee.setRef(schema.name());
+        }
```
Comment (jbee) [inline appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: I'd prefer a single `referee.setRef(...)` with a conditional expression but that is personal preference.

Comment (Pscheidl) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: Why not ? Can be squeezed into a single line quite easily. 

```
referee.setRef(schema == null || schema.name().isEmpty() ? referenceClass.getSimpleName() : schema.name());
```

Changed in https://github.com/payara/Payara/pull/4494/commits/2e2e6f6f7a85351e4243eaf6fbf2b6b9248a5bc2.

File: appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java (hunk 3 above)
Comment (Pscheidl) [reply appserver/payara-appserver-modules/microprofile/openapi/src/test/java/fish/payara/microprofile/openapi/test/app/application/OASchemaReferenceNameTest.java]: Done.

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java (hunk 4 above)
Comment (jbee) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: Don't know why, some people do not like conditionals.

Comment (rdebusscher) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: ternary operators make it far less readable, compiler can do this optimization.

Comment (jbee) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: > ternary operators make it far less readable

That is subjective. I can read conditions "better".

> compiler can do this optimization.

Its not a readability or performance optimization as far as I am concerned but a semantic one.  With the conditional it is very clear that we want to do this (here `setRef`) with some value in any branch. This is much clearer expressed by not doing that in an `if-else` where both branches do it but a single call with a argument that is either this or that.

Comment (Pscheidl) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java]: I'm fine with both. I can read almost any code of any style. Feel free to adjust it to your liking.

**Comment by jbee** [comment]
jenkins test please

**Pull request merged**